### DISCUSSION TITLE
How to configure Ocelot to prevent it from adding the port to my downstream URL?

### DISCUSSION THREAD

**Discussion by walissonde** [ThreeMammals/Ocelot#1750] [Q&A]
Hello,
I am currently working on a Proof of Concept (POC) using Ocelot, but I have encountered an issue when making a URL call from localhost to route traffic to my API.
**Is that correct when missing downstream port is replaced with an upstream port?**

## Actual Behavior 
The base URL of my project is:
- Host: **localhost**
- Port: **5021**

I also have a downstream URL, for instance:
- Downstream URL: `my-external-api.com` (without specifying any port)

However, when I make a call to my URL through the Ocelot gateway, it adds a port to my URL, causing my external API to malfunction because it doesn't expect a port.

Here's what happens:
> I call my upstream URL as follows: `localhost:5021/gateway/my-api`
> Ocelot attempts to call my downstream and adds a port: `my-external-api.com:5021`

## Expected Behavior 
Now, I'm looking for a way to configure Ocelot to ignore this added port (e.g., 5021) when making calls to my API. I've already removed the port from my downstream configuration.

**How can I configure Ocelot to resolve this issue and prevent it from adding the port to my downstream URL?**

## Specifications
  - Version: **19.0.2**
  - Platform: **.NET 7**
  - Subsystem: Windows

**Comment by raman-m** [comment]
Hi Walisson!
Thanks for your interest in Ocelot!

As far as I understood, *Ocelot substitutes downstream port by upstream port when there is no downstream port.* Correct?
If Yes, it can be a bug. But this scenario should be reproduced and proved.
Also, you need to check Ocelot logs for a warnings & messages of configuration validation.
Could you check your logs for messages please?

On my opinion, when there is no specified downstream port, so a warning should be logged.
Moreover, as a default feature, missing port must be initialized by default port of current scheme (protocol) like:
- http: 80
- https: 443
- ws: ?
- etc.

**Comment by raman-m** [comment]
> How can I configure Ocelot to resolve this issue and prevent it from adding the port to my downstream URL?

Just define the downstream port for all routes! For example:
```json
{
  "Routes": [
    {
      // Route 1
      "DownstreamScheme": "http",
      "DownstreamHostAndPorts": [
        {
          "Host": "route1.com",
          "Port": 80
        }
      ],
      // ...
    },
    {
      // Route 2
      "DownstreamScheme": "https",
      "DownstreamHostAndPorts": [
        {
          "Host": "route2.com",
          "Port": 443
        }
      ],
      // ...
    },
  ],
  "GlobalConfiguration": {
  }
}
```

Let me know testing results please!

**Comment by raman-m** [comment]
> @walissonde commented [on Sep 13](#discussion-5754916)

Answering your question
> How can I configure Ocelot to resolve this issue and prevent it from adding the port to my downstream URL?

I [suggested you](#discussioncomment-7331495) just specify Port value explicitly as is done for all routes now as the default behavior. 

Regarding potential bug of port substitution from upstream... Still need to check it... Haven't you received validation errors 7 warning in the log?

Also, I've looked into the code, the [DownstreamRequestCreator](/search?q=repo%3AThreeMammals%2FOcelot%20DownstreamRequestCreator&type=code) class, and it has no extra logic related to `Port` property.
The [Port](/search?q=repo%3AThreeMammals%2FOcelot+Port+language%3AC%23+path%3A%2F%5Esrc%5C%2F%2F&type=code) property has no implicit initializations...
Historically Port should be initialized explicitly in configuration of the Route, because downstream services have deployed behind firewalls which open access with specific ports.

----
**Have I answered on your question?
Would you like to convert your question to bug/feature type?**

**Comment by raman-m** [comment]
@walissonde  It could be a bug, it can be a feature, I'm not sure now. Historically all routes should have specified port.

**Reply by walissonde** [reply]
Thank you, Raman. I apologize for my delayed response. I believe creating a feature in Ocelot that allows downstream endpoints to operate without a mandatory port could be incredibly valuable. Picture this scenario: you want to run your API gateway locally and connect to an external API without needing to specify a port. I currently find myself in this exact situation where none of my downstream endpoints require a port. However, when running locally, my localhost automatically has a port. Strangely, Ocelot inserts my localhost port into the downstream URL when I call this endpoint via the Ocelot gateway.

**Reply by ggnaegi** [reply]
@walissonde, I agree with @raman-m, the port must be specified. Only if the port is 80 (http) or 443 (https), then the call to the api will be made "without - I mean it's not part of the url -" a port number because 80 is the default for http and 443 the default for https.

**Reply by raman-m** [reply]
Agreed? Are we going to create a feature-issue to implement default ports for common schemes /protocols?

Note, ror Websockets there is no default port, so `ws` and `wss` schemes appropriated routes must have ports.

**Reply by raman-m** [reply]
@walissonde Please, chose the answer, and if you have extra time, create issue which I accept. Use this discussion to write description.
